File: app/src/main/java/com/esafirm/androidplayground/utils/Logger.java
package com.esafirm.androidplayground.utils;

import android.content.*;
import android.graphics.*;
import android.util.*;
import android.view.*;
import android.widget.*;

import rx.*;
import rx.functions.*;
import rx.subjects.*;

@SuppressWarnings("unchecked")
public class Logger {

    private static SerializedSubject subject = new SerializedSubject<>(PublishSubject.create());
    private static final StringBuilder loggerBuilder = new StringBuilder();

    public static void clear() {
        loggerBuilder.setLength(0);
    }

    public static void divider() {
        log("-----------");
    }

    public static void log(Object o) {
        String text = o != null ? o.toString() : "Object null";
        System.out.println(text);
        loggerBuilder.append(text).append("\n");
        subject.onNext(loggerBuilder.toString());
    }

    public static String getLogText() {
        return loggerBuilder.toString();
    }

    private static Observable<String> getObs() {
        return subject.startWith(getLogText()).onBackpressureLatest();
    }

    public static ViewGroup getLogView(Context context) {
        int padding = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 16, context.getResources().getDisplayMetrics());
        final TextView textView = new TextView(context);
        textView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        textView.setTextColor(Color.BLACK);
        textView.setPadding(padding, padding / 2, padding, padding / 2);

        ScrollView scrollView = new ScrollView(context);
        scrollView.addView(textView);

        getObs().observeOn(rx.android.schedulers.AndroidSchedulers.mainThread())
                .subscribe(new Action1<String>() {
                    @Override
                    public void call(String s) {
                        textView.setText(null);
                        textView.setText(s);
                    }
                });

        return scrollView;
    }
}

File: app/src/main/java/com/esafirm/androidplayground/utils/Logger.kt
package com.esafirm.androidplayground.utils

import android.content.Context
import android.graphics.Color
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.ScrollView
import android.widget.TextView
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.processors.PublishProcessor

object Logger {

    private val subject = PublishProcessor.create<String>()
    private val loggerBuilder = StringBuilder()

    private val logText: String
        get() = loggerBuilder.toString()

    private val obs: Flowable<String>
        get() = subject.startWith(logText).onBackpressureLatest()

    fun clear() {
        loggerBuilder.setLength(0)
    }

    fun divider() {
        log("-----------")
    }

    fun log(o: Any?) {
        val text = o?.toString() ?: "Object null"
        println(text)
        loggerBuilder.append(text).append("\n")
        subject.onNext(loggerBuilder.toString())
    }

    fun getLogView(context: Context): ViewGroup {
        val padding = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, 16f, context.resources.displayMetrics
        ).toInt()

        val textView = TextView(context)
        textView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        textView.setTextColor(Color.BLACK)
        textView.setPadding(padding, padding / 2, padding, padding / 2)

        var disposable: Disposable? = null

        textView.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
            override fun onViewDetachedFromWindow(p0: View?) {
                disposable?.dispose()
            }

            override fun onViewAttachedToWindow(p0: View?) {
                disposable?.dispose()
                disposable = obs.observeOn(AndroidSchedulers.mainThread())
                    .subscribe {
                        textView.text = null
                        textView.text = it
                    }

            }
        })

        val scrollView = ScrollView(context)
        scrollView.addView(textView)
        return scrollView
    }
}
